### PR TITLE
Add guide for csi deployment

### PR DESCRIPTION
**Pull request by hyoung-90**

### CODE REVIEW

**Comment by hyoung-90** [comment]
It's ready for the final review.

File: deploy/csi/kms-config.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        "vaultAuthPath": "/v1/auth/kubernetes/login",
+        "vaultRole": "csi-kubernetes",
+        "vaultPassphraseRoot": "/v1/secret",
+        "vaultPassphrasePath": "ceph-csi/",
+        "vaultCAVerify": "false"
+      }
+    }
+metadata:
+  name: ceph-csi-encryption-kms-config
+  namespace: ceph-csi
```
Comment (hookak) [inline deploy/csi/kms-config.yaml:19]: why do we need this file?

File: deploy/csi/rbd/pvc.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+---
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: rbd-pvc
+spec:
+  accessModes:
+    - ReadWriteOnce
+  volumeMode: Block
```
Comment (hookak) [inline deploy/csi/rbd/pvc.yaml:9]: I have `losetup -j` issue when mounting `volumeMode: Block pvc`
I deployed minikube using 'virtualbox'

File: docs/USERGUIDE.md [hunk 3]
```diff
@@ -57,3 +57,188 @@ $ make clean
...
+> Note: This part will be added soon.
+
+### 2. Deployment with K8s
+#### Deploy namespace
+```shell script
+# Namespace where the csi plugin pods and other resources are created
+$ kubectl apply -f deploy/csi/namespace.yaml
+```
+#### Deploy config
+Modify the value of `clusterID` and `monitors` in `deploy/csi/csi-config-map.yaml`.
```
Comment (hookak) [inline docs/USERGUIDE.md:115]: Could we explain more detail with yaml example?

**Review by hookak — CHANGES_REQUESTED** [review]

**Review by hbinkim — COMMENTED** [review]
Some parts that I wasn't sure about...

File: docs/USERGUIDE.md [hunk 4]
```diff
@@ -57,3 +57,188 @@ $ make clean
...
+
+|Parameter |Value |
+|---|---|
+|`clusterID` | `fsid` from the step 1-1 |
+|`monitors` | `mon_host` from the step 1-1 |
+```shell script
+$ kubectl apply -f deploy/csi/csi-config-map.yaml
```
[the remark this answers is not in the document]
Comment (hookak) [reply docs/USERGUIDE.md:124]: Setting a subvolumegroup is optional. In ceph csi, subvolume will be created in default subvolumegroup(csi) and, if subvolumegroup didn't exist, ceph csi will create the default subvolumegroup(csi).

File: docs/USERGUIDE.md [hunk 5]
```diff
@@ -57,3 +57,190 @@ $ make clean
...
 $ make uninstall
 ```
+
+## Use external ceph cluster in Kubernetes
+This is a guide for connecting external ceph cluster to k8s using [Ceph CSI](https://github.com/ceph/ceph-csi). Ceph CSI plugins implement an interface between Kubernetes and Ceph cluster.
+Independent CSI plugins are provided to support RBD and CephFS backed volumes. The Ceph CSI version used in this guide is [v3.1.2](https://github.com/ceph/ceph-csi/releases/tag/v3.1.2).
+
+### 1. Get information of ceph cluster
+
+#### Install ceph cluster separately
```
Comment (hbinkim) [inline docs/USERGUIDE.md:67]: Ceph cluster installed without HyperSDS-Operator (ex. cephadm or ansible)

seems more clear

File: docs/USERGUIDE.md [hunk 6]
```diff
@@ -57,3 +57,190 @@ $ make clean
...
+```shell script
+$ ceph fs volume ls
+[
+    {
+        "name": "myfs"
+    }
+]
+```
+
+#### Install ceph cluster using HyperSDS-Operator
```
Comment (hbinkim) [inline docs/USERGUIDE.md:107]: Ceph cluster installed with HyperSDS-Operator

seems more clear

File: docs/USERGUIDE.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (hyoung-90) [reply docs/USERGUIDE.md:124]: `cephFS.subvolumeGroup` will be deleted.

**Review by hbinkim — COMMENTED** [review]
Just a minor expression detail

File: deploy/csi/cephfs/secret.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+---
+apiVersion: v1
+kind: Secret
+metadata:
+  name: csi-cephfs-secret
+  namespace: ceph-csi
+stringData:
+  # Required for statically provisioned volumes
+  userID: mds.myfs.s1-2.esxjmt
```
[the remark this answers is not in the document]
Comment (hyoung-90) [reply deploy/csi/cephfs/secret.yaml:9]: This is required for static provisioning. Users do not need to modify them unless static is required.
I will add additional comment.

File: docs/USERGUIDE.md [hunk 8]
```diff
@@ -57,3 +57,190 @@ $ make clean
 # Clean only CRD
 $ make uninstall
 ```
+
```
Comment (hbinkim) [inline docs/USERGUIDE.md:60]: Maybe adding a table of contents at the top is helpful

``` markdown
## Table of Contents

- [Install Operator](##Install-Hypersds-Operator)
- [Bootstrap Ceph Cluster](##Use-Hypersds-Operator)
- [Uninstall Operator](##Uninstall-Hypersds-Operator)
- [Connect Ceph to K8s](##Use-external-ceph-cluster-in-Kubernetes)
```

Comment (hbinkim) [reply docs/USERGUIDE.md:67]: @hyoung-90 sorry my bad if you remove one #, it works

``` markdown
## Table of Contents

- [Install Operator](#Install-Hypersds-Operator)
- [Bootstrap Ceph Cluster](#Use-Hypersds-Operator)
- [Uninstall Operator](#Uninstall-Hypersds-Operator)
- [Connect Ceph to K8s](#Use-external-ceph-cluster-in-Kubernetes)
```

File: deploy/csi/kms-config.yaml (hunk 1 above)
Comment (hbinkim) [reply deploy/csi/kms-config.yaml:19]: I'm not fully aware of the details but as far as I know, ksm (key management system) is needed to generate a passphrase for each encrypted rbd image. 

https://github.com/nixpanic/ceph-csi/blob/61c56576dd441aedbd954148982b585c467972cf/docs/deploy-rbd.md#encryption-kms-configuration

File: docs/USERGUIDE.md (hunk 3 above)
Comment (hyoung-90) [reply docs/USERGUIDE.md]: I'll add a comment to the values that need to be modified.
`# UPDATE THIS VALUE`
`adminKey: AQCeBcZftAEvExAAultsKBpNpiWWGi06Md7mmw==`

File: deploy/csi/kms-config.yaml (hunk 1 above)
Comment (hyoung-90) [reply deploy/csi/kms-config.yaml:19]: Currently we don't set or use encryption for rbd image, so we don't need it.

File: deploy/csi/rbd/pvc.yaml (hunk 2 above)
Comment (hyoung-90) [reply deploy/csi/rbd/pvc.yaml:9]: In most cases, `rbd` is used for `block`, so it seems that `volumeMode` should be `Block`.
`losetup -j` issue seems to be caused by the virtualbox environment.

File: deploy/csi/rbd/csi-rbdplugin-provisioner.yaml [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+          imagePullPolicy: "IfNotPresent"
+          volumeMounts:
+            - name: socket-dir
+              mountPath: /csi
+        - name: csi-rbdplugin
+          securityContext:
+            privileged: true
+            capabilities:
+              add: ["SYS_ADMIN"]
+          # for stable functionality replace canary with latest release version
```
Comment (hbinkim) [inline deploy/csi/rbd/csi-rbdplugin-provisioner.yaml:103]: Unnecessary comment found

Comment (hbinkim) [inline deploy/csi/rbd/csi-rbdplugin-provisioner.yaml:103]: Unnecessary comment found

File: deploy/csi/cephfs/csi-cephfsplugin-provisioner.yaml [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+          imagePullPolicy: "IfNotPresent"
+          volumeMounts:
+            - name: socket-dir
+              mountPath: /csi
+        - name: csi-cephfsplugin
+          securityContext:
+            privileged: true
+            capabilities:
+              add: ["SYS_ADMIN"]
+          # for stable functionality replace canary with latest release version
```
Comment (hbinkim) [inline deploy/csi/cephfs/csi-cephfsplugin-provisioner.yaml:103]: Unnecessary comment found

**Review by hbinkim — COMMENTED** [review]
A few more minor details

**Comment by hbinkim** [comment]
LGTM

**Review by hookak — APPROVED** [review]
lgtm

**Pull request merged**